crates/hir_def/src/lib.rs

```diff
@@ -602,24 +602,27 @@ impl AsMacroCall for InFile<&ast::MacroCall> {
         krate: CrateId,
         resolver: impl Fn(path::ModPath) -> Option<MacroDefId>,
         error_sink: &mut dyn FnMut(mbe::ExpandError),
-    ) -> Option<MacroCallId> {
+    ) -> Result<Option<MacroCallId>, UnresolvedMacro> {
```

Comment:
Is is possible it just return a `Result<MacroCallId>` here ?

Reply:
With the current structure, no. All 3 variants represent unique outcomes and each outcome needs to be handled differently by the caller.

But, as you alluded to in your other comment, you can encode the UnresolvedMacro variant through the error sink. In that case, you could just leave the return type as `Option<MacroCallId>`.

The `Err(UnresolvedMacro)` case can be identified by checking for `None` + no expand error written to sink.
The `Ok(None)` case is now `None` + error written to sink.
`Ok(Some(id))` is now `Some(id)`.

I decided to take the approach I did to be consistent with `macro_call_as_call_id` and the overall direction of https://github.com/rust-analyzer/rust-analyzer/pull/7805 with the introduction of UnresolvedMacro as well as wanting to avoid the caller having to make the assumption that no expand error implies an unresolved macro. I have no problems changing it though.

Reply:
> I decided to take the approach I did to be consistent with macro_call_as_call_id and the overall direction of #7805 with the introduction of UnresolvedMacro as well as wanting to avoid the caller having to make the assumption that no expand error implies an unresolved macro. I have no problems changing it though

Sound reasonable ! 

And according the same logic, do you think changing it to `Result<Result<MacroCallId, ErrorEmitted>, UnresolvedMacro>` directly,  would be more suitable ? `Ok(None)`( for `Result<Option<MacroCallId>, UnresolvedMacro>`) imples that there are some errors, and because the direction of #7805 is make these kind of thing explicit, so It will make it more explicit too.

Reply:
Yes, that sounds like a good idea.